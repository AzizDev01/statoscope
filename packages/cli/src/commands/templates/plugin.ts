import { concatTemplateParts, templatePartsByOptions } from './templateParts';
import { FileExt, RuleTemplatesPartsFromFileExt, Template } from './types';

export const pluginTemplate: Template = (templateOptions) => {
  const plugin = templatePartsByOptions(templateOptions, {
    js: jsPlugin,
    ts: tsPlugin,
  });

  return concatTemplateParts(plugin);
};

const jsPlugin: RuleTemplatesPartsFromFileExt[FileExt.js] = {
  imports: {
    commonjs: `const asyncChunkRule = require('./rule.statoscope.js')`,
    esm: `import asyncChunkRule from './rule.statoscope.js';`,
  },
  body: `/**
  * @typedef {import('@statoscope/webpack-model').Prepared} Prepared
  * @returns {import('@statoscope/stats-validator/dist/plugin').PluginDescription<Prepared>} PluginDescription
 */
const newStatoscopePlugin = () => {
  return {
    rules: {
      "async-chunk": asyncChunkRule,
    },
  };
};`,
  exports: {
    commonjs: 'module.exports = newStatoscopePlugin',
    esm: 'export default newStatoscopePlugin',
  },
};

const tsPlugin: RuleTemplatesPartsFromFileExt[FileExt.ts] = {
  imports: `import type { PluginFn } from '@statoscope/stats-validator/dist/plugin';
import type { Prepared } from '@statoscope/webpack-model';
import asyncChunkRule from './rule.statoscope.ts';`,
  body: `
const newStatoscopePlugin: PluginFn<Prepared> = () => {
  return {
    rules: {
      "async-chunk": asyncChunkRule,
    },
  };
};`,
  exports: 'export default newStatoscopePlugin',
};
