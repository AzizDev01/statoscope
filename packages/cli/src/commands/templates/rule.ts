export function ruleTemplate(): string {
  return `import { Rule } from '@statoscope/stats-validator/dist/rule';
import { Prepared } from '@statoscope/webpack-model';

//Statoscope Plugins examples:
https://github.com/statoscope/statoscope/tree/c9d0c148c1e38a87696d79485f37e598110db80c/packages/stats-validator-plugin-webpack#rules


${asyncChunkRule}

export default asyncChunkRule;`;
}

export const asyncChunkRule = `
type WebpackRule<TParams> = Rule<TParams, Prepared>;

type Params = string[];

const asyncChunkRule: WebpackRule<Params> = (ruleParams, data, api): void => {
  const normalizedParams: string[] = ruleParams ? ruleParams : [];
  const chunks = Array.from(new Set(...data.compilations.map((c) => c.chunks)));
  chunks
    .filter((c) =>
      c.names.some((n) => normalizedParams.includes(n) && c.initial !== false)
    )
    .forEach((f) => {
      api.message(f.names[0] + ' is not async chunk.');
    });
};`;
