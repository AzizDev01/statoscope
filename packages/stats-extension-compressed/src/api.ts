import {
  default as makeEntityResolver,
  Resolver,
} from '@statoscope/helpers/dist/entity-resolver';
import { APIFactory } from '@statoscope/extensions';
import { Format, Resource, Size } from './generator';

export type API = (compilationId: string, resourceId: string) => Size | null;

const makeAPI: APIFactory<Format, API> = (source) => {
  const sizeResolvers: Map<string, Resolver<string, Resource>> = new Map();

  for (const compilation of source.payload.compilations) {
    sizeResolvers.set(
      compilation.id,
      makeEntityResolver(compilation.resources, (r) => r.id)
    );
  }

  return (compilationId: string, resourceId: string): Size | null => {
    return sizeResolvers.get(compilationId)?.(resourceId)?.size ?? null;
  };
};

export default makeAPI;
