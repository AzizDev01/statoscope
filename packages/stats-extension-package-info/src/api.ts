import {
  default as makeEntityResolver,
  Resolver,
} from '@statoscope/helpers/dist/entity-resolver';
import { APIFactory } from '@statoscope/extensions';
import { Format, Package, Instance } from './generator';

export type API = {
  getPackage: (compilationId: string, packageName: string) => Package | null;
  getInstance: (
    compilationId: string,
    packageName: string,
    instancePath: string
  ) => Instance | null;
};

const makeAPI: APIFactory<Format, API> = (source) => {
  const packageResolvers: Map<string, Resolver<string, Package>> = new Map();
  const instanceResolvers: Map<Package, Resolver<string, Instance>> = new Map();

  for (const compilation of source.payload.compilations) {
    packageResolvers.set(
      compilation.id,
      makeEntityResolver(compilation.packages, (r) => r.name)
    );

    for (const packageItem of compilation.packages) {
      instanceResolvers.set(
        packageItem,
        makeEntityResolver(packageItem.instances, (r) => r.path)
      );
    }
  }

  return {
    getPackage(compilationId: string, packageId: string): Package | null {
      return packageResolvers.get(compilationId)?.(packageId) ?? null;
    },
    getInstance(
      compilationId: string,
      packageId: string,
      instancePath: string
    ): Instance | null {
      const resolvedPackage = packageResolvers.get(compilationId)?.(packageId) ?? null;

      if (!resolvedPackage) {
        return null;
      }

      return instanceResolvers.get(resolvedPackage)?.(instancePath) ?? null;
    },
  };
};

export default makeAPI;
