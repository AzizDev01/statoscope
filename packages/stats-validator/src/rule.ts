import { API } from './api';

export type RuleDataInput<TInput> = { reference?: TInput; input: TInput };

export type Rule<TParams, TInput> = (
  params: TParams | null,
  data: RuleDataInput<TInput>,
  api: API
) => void;
