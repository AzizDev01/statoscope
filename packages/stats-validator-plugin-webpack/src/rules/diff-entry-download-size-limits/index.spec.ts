import { makeAPI } from '@statoscope/stats-validator/dist/api';
import { RuleDataInput } from '@statoscope/stats-validator/dist/rule';
import { Prepared } from '@statoscope/webpack-model';
import { API } from '@statoscope/types/types/validation/api';
import plugin from '../..';
import referenceStats from '../../../../../test/bundles/simple/stats-prod.json';
import afterStats from '../../../../../test/bundles/simple/stats-dev.json';
import rule, { Params } from './';

function prepareAPI(ruleParams: Params): API {
  const pluginInstance = plugin();
  const prepared: RuleDataInput<Prepared> = pluginInstance.prepare!([
    { name: 'input.json', data: afterStats },
    { name: 'reference.json', data: referenceStats },
  ]);
  const api = makeAPI();

  rule(ruleParams, prepared, api);

  return api;
}

type Case = {
  name: string;
  params: Params;
};

const maxSizeCases: Case[] = [
  {
    name: 'global',
    params: {
      global: {
        maxSizeDiff: 100,
      },
    },
  },
  {
    name: 'override global',
    params: {
      global: {
        maxSizeDiff: 100000,
      },

      byName: [
        {
          name: /.+/,
          limits: {
            maxSizeDiff: 100,
          },
        },
      ],
    },
  },
  {
    name: 'not match - global',
    params: {
      global: {
        maxSizeDiff: 100000,
      },
    },
  },
  {
    name: 'not match - override global',
    params: {
      global: {
        maxSizeDiff: 100,
      },

      byName: [
        {
          name: /.+/,
          limits: {
            maxSizeDiff: 100000,
          },
        },
      ],
    },
  },
  {
    name: 'with useCompressedSize = false',
    params: {
      useCompressedSize: false,
      global: {
        maxSizeDiff: 100,
      },
    },
  },
];

const maxInitialSizeCases: Case[] = [
  {
    name: 'global',
    params: {
      global: {
        maxInitialSizeDiff: 100,
      },
    },
  },
  {
    name: 'override global',
    params: {
      global: {
        maxInitialSizeDiff: 100000,
      },

      byName: [
        {
          name: /.+/,
          limits: {
            maxInitialSizeDiff: 100,
          },
        },
      ],
    },
  },
  {
    name: 'not match - global',
    params: {
      global: {
        maxInitialSizeDiff: 100000,
      },
    },
  },
  {
    name: 'not match - override global',
    params: {
      global: {
        maxInitialSizeDiff: 100,
      },

      byName: [
        {
          name: /.+/,
          limits: {
            maxInitialSizeDiff: 100000,
          },
        },
      ],
    },
  },
  {
    name: 'with useCompressedSize = false',
    params: {
      useCompressedSize: false,
      global: {
        maxInitialSizeDiff: 100,
      },
    },
  },
];

const maxAsyncSizeCases: Case[] = [
  {
    name: 'global',
    params: {
      global: {
        maxAsyncSizeDiff: 100,
      },
    },
  },
  {
    name: 'override global',
    params: {
      global: {
        maxAsyncSizeDiff: 100000,
      },

      byName: [
        {
          name: /.+/,
          limits: {
            maxAsyncSizeDiff: 100,
          },
        },
      ],
    },
  },
  {
    name: 'not match - global',
    params: {
      global: {
        maxAsyncSizeDiff: 100000,
      },
    },
  },
  {
    name: 'not match - override global',
    params: {
      global: {
        maxAsyncSizeDiff: 100,
      },

      byName: [
        {
          name: /.+/,
          limits: {
            maxAsyncSizeDiff: 100000,
          },
        },
      ],
    },
  },
  {
    name: 'with useCompressedSize = false',
    params: {
      useCompressedSize: false,
      global: {
        maxAsyncSizeDiff: 100,
      },
    },
  },
];

describe('maxSizeDiff', () => {
  test.each(maxSizeCases)('$name', (item) => {
    const api = prepareAPI(item.params);
    expect(api.getStorage()).toMatchSnapshot();
  });
});

describe('maxInitialSizeDiff', () => {
  test.each(maxInitialSizeCases)('$name', (item) => {
    const api = prepareAPI(item.params);
    expect(api.getStorage()).toMatchSnapshot();
  });
});

describe('maxAsyncSizeDiff', () => {
  test.each(maxAsyncSizeCases)('$name', (item) => {
    const api = prepareAPI(item.params);
    expect(api.getStorage()).toMatchSnapshot();
  });
});
