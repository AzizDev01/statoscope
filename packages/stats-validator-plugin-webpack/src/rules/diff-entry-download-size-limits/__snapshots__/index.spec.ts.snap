// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`maxAsyncSizeDiff global 1`] = `
Array [
  Object {
    "compilation": "9eb9a218218f1ee4ba40",
    "details": Array [
      Object {
        "payload": Object {
          "context": Object {
            "after": Object {
              "asyncSize": 590,
              "initialSize": 12211,
              "size": 12801,
            },
            "before": Object {
              "asyncSize": 134,
              "initialSize": 8615,
              "size": 8749,
            },
            "compilation": "9eb9a218218f1ee4ba40",
            "diff": Object {
              "asyncSize": Object {
                "absolute": 456,
                "ok": false,
                "percent": 340.2985074626866,
              },
              "initialSize": Object {
                "absolute": 3596,
                "ok": true,
                "percent": 41.74114915844458,
              },
              "size": Object {
                "absolute": 4052,
                "ok": true,
                "percent": 46.313864441650466,
              },
            },
            "entry": "two",
            "rule": Object {
              "maxAsyncSizeDiff": 100,
              "maxInitialSizeDiff": Infinity,
              "maxSizeDiff": Infinity,
            },
            "useCompressedSize": true,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              before: #.before,
              after: #.after,
              diff: #.diff,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"two\\": Download size diff of async assets is 0.45 kb (340.30%). It's over the 0.10 kb limit",
    "related": Array [
      Object {
        "id": "two",
        "type": "entry",
      },
    ],
  },
]
`;

exports[`maxAsyncSizeDiff not match - global 1`] = `Array []`;

exports[`maxAsyncSizeDiff not match - override global 1`] = `Array []`;

exports[`maxAsyncSizeDiff override global 1`] = `
Array [
  Object {
    "compilation": "9eb9a218218f1ee4ba40",
    "details": Array [
      Object {
        "payload": Object {
          "context": Object {
            "after": Object {
              "asyncSize": 590,
              "initialSize": 12211,
              "size": 12801,
            },
            "before": Object {
              "asyncSize": 134,
              "initialSize": 8615,
              "size": 8749,
            },
            "compilation": "9eb9a218218f1ee4ba40",
            "diff": Object {
              "asyncSize": Object {
                "absolute": 456,
                "ok": false,
                "percent": 340.2985074626866,
              },
              "initialSize": Object {
                "absolute": 3596,
                "ok": true,
                "percent": 41.74114915844458,
              },
              "size": Object {
                "absolute": 4052,
                "ok": true,
                "percent": 46.313864441650466,
              },
            },
            "entry": "two",
            "rule": Object {
              "maxAsyncSizeDiff": 100,
              "maxInitialSizeDiff": Infinity,
              "maxSizeDiff": Infinity,
            },
            "useCompressedSize": true,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              before: #.before,
              after: #.after,
              diff: #.diff,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"two\\": Download size diff of async assets is 0.45 kb (340.30%). It's over the 0.10 kb limit",
    "related": Array [
      Object {
        "id": "two",
        "type": "entry",
      },
    ],
  },
]
`;

exports[`maxAsyncSizeDiff with useCompressedSize = false 1`] = `
Array [
  Object {
    "compilation": "9eb9a218218f1ee4ba40",
    "details": Array [
      Object {
        "payload": Object {
          "context": Object {
            "after": Object {
              "asyncSize": 1200,
              "initialSize": 29129,
              "size": 30329,
            },
            "before": Object {
              "asyncSize": 142,
              "initialSize": 12277,
              "size": 12419,
            },
            "compilation": "9eb9a218218f1ee4ba40",
            "diff": Object {
              "asyncSize": Object {
                "absolute": 1058,
                "ok": false,
                "percent": 745.0704225352113,
              },
              "initialSize": Object {
                "absolute": 16852,
                "ok": true,
                "percent": 137.26480410523743,
              },
              "size": Object {
                "absolute": 17910,
                "ok": true,
                "percent": 144.21451002496175,
              },
            },
            "entry": "two",
            "rule": Object {
              "maxAsyncSizeDiff": 100,
              "maxInitialSizeDiff": Infinity,
              "maxSizeDiff": Infinity,
            },
            "useCompressedSize": false,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              before: #.before,
              after: #.after,
              diff: #.diff,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"two\\": Download size diff of async assets is 1.03 kb (745.07%). It's over the 0.10 kb limit",
    "related": Array [
      Object {
        "id": "two",
        "type": "entry",
      },
    ],
  },
]
`;

exports[`maxInitialSizeDiff global 1`] = `
Array [
  Object {
    "compilation": "9eb9a218218f1ee4ba40",
    "details": Array [
      Object {
        "payload": Object {
          "context": Object {
            "after": Object {
              "asyncSize": 0,
              "initialSize": 1887,
              "size": 1887,
            },
            "before": Object {
              "asyncSize": 0,
              "initialSize": 449,
              "size": 449,
            },
            "compilation": "9eb9a218218f1ee4ba40",
            "diff": Object {
              "asyncSize": Object {
                "absolute": 0,
                "ok": true,
                "percent": 0,
              },
              "initialSize": Object {
                "absolute": 1438,
                "ok": false,
                "percent": 320.26726057906455,
              },
              "size": Object {
                "absolute": 1438,
                "ok": true,
                "percent": 320.26726057906455,
              },
            },
            "entry": "one",
            "rule": Object {
              "maxAsyncSizeDiff": Infinity,
              "maxInitialSizeDiff": 100,
              "maxSizeDiff": Infinity,
            },
            "useCompressedSize": true,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              before: #.before,
              after: #.after,
              diff: #.diff,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"one\\": Download size diff of initial assets is 1.40 kb (320.27%). It's over the 0.10 kb limit",
    "related": Array [
      Object {
        "id": "one",
        "type": "entry",
      },
    ],
  },
  Object {
    "compilation": "9eb9a218218f1ee4ba40",
    "details": Array [
      Object {
        "payload": Object {
          "context": Object {
            "after": Object {
              "asyncSize": 590,
              "initialSize": 12211,
              "size": 12801,
            },
            "before": Object {
              "asyncSize": 134,
              "initialSize": 8615,
              "size": 8749,
            },
            "compilation": "9eb9a218218f1ee4ba40",
            "diff": Object {
              "asyncSize": Object {
                "absolute": 456,
                "ok": true,
                "percent": 340.2985074626866,
              },
              "initialSize": Object {
                "absolute": 3596,
                "ok": false,
                "percent": 41.74114915844458,
              },
              "size": Object {
                "absolute": 4052,
                "ok": true,
                "percent": 46.313864441650466,
              },
            },
            "entry": "two",
            "rule": Object {
              "maxAsyncSizeDiff": Infinity,
              "maxInitialSizeDiff": 100,
              "maxSizeDiff": Infinity,
            },
            "useCompressedSize": true,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              before: #.before,
              after: #.after,
              diff: #.diff,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"two\\": Download size diff of initial assets is 3.51 kb (41.74%). It's over the 0.10 kb limit",
    "related": Array [
      Object {
        "id": "two",
        "type": "entry",
      },
    ],
  },
]
`;

exports[`maxInitialSizeDiff not match - global 1`] = `Array []`;

exports[`maxInitialSizeDiff not match - override global 1`] = `Array []`;

exports[`maxInitialSizeDiff override global 1`] = `
Array [
  Object {
    "compilation": "9eb9a218218f1ee4ba40",
    "details": Array [
      Object {
        "payload": Object {
          "context": Object {
            "after": Object {
              "asyncSize": 0,
              "initialSize": 1887,
              "size": 1887,
            },
            "before": Object {
              "asyncSize": 0,
              "initialSize": 449,
              "size": 449,
            },
            "compilation": "9eb9a218218f1ee4ba40",
            "diff": Object {
              "asyncSize": Object {
                "absolute": 0,
                "ok": true,
                "percent": 0,
              },
              "initialSize": Object {
                "absolute": 1438,
                "ok": false,
                "percent": 320.26726057906455,
              },
              "size": Object {
                "absolute": 1438,
                "ok": true,
                "percent": 320.26726057906455,
              },
            },
            "entry": "one",
            "rule": Object {
              "maxAsyncSizeDiff": Infinity,
              "maxInitialSizeDiff": 100,
              "maxSizeDiff": Infinity,
            },
            "useCompressedSize": true,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              before: #.before,
              after: #.after,
              diff: #.diff,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"one\\": Download size diff of initial assets is 1.40 kb (320.27%). It's over the 0.10 kb limit",
    "related": Array [
      Object {
        "id": "one",
        "type": "entry",
      },
    ],
  },
  Object {
    "compilation": "9eb9a218218f1ee4ba40",
    "details": Array [
      Object {
        "payload": Object {
          "context": Object {
            "after": Object {
              "asyncSize": 590,
              "initialSize": 12211,
              "size": 12801,
            },
            "before": Object {
              "asyncSize": 134,
              "initialSize": 8615,
              "size": 8749,
            },
            "compilation": "9eb9a218218f1ee4ba40",
            "diff": Object {
              "asyncSize": Object {
                "absolute": 456,
                "ok": true,
                "percent": 340.2985074626866,
              },
              "initialSize": Object {
                "absolute": 3596,
                "ok": false,
                "percent": 41.74114915844458,
              },
              "size": Object {
                "absolute": 4052,
                "ok": true,
                "percent": 46.313864441650466,
              },
            },
            "entry": "two",
            "rule": Object {
              "maxAsyncSizeDiff": Infinity,
              "maxInitialSizeDiff": 100,
              "maxSizeDiff": Infinity,
            },
            "useCompressedSize": true,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              before: #.before,
              after: #.after,
              diff: #.diff,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"two\\": Download size diff of initial assets is 3.51 kb (41.74%). It's over the 0.10 kb limit",
    "related": Array [
      Object {
        "id": "two",
        "type": "entry",
      },
    ],
  },
]
`;

exports[`maxInitialSizeDiff with useCompressedSize = false 1`] = `
Array [
  Object {
    "compilation": "9eb9a218218f1ee4ba40",
    "details": Array [
      Object {
        "payload": Object {
          "context": Object {
            "after": Object {
              "asyncSize": 0,
              "initialSize": 8348,
              "size": 8348,
            },
            "before": Object {
              "asyncSize": 0,
              "initialSize": 799,
              "size": 799,
            },
            "compilation": "9eb9a218218f1ee4ba40",
            "diff": Object {
              "asyncSize": Object {
                "absolute": 0,
                "ok": true,
                "percent": 0,
              },
              "initialSize": Object {
                "absolute": 7549,
                "ok": false,
                "percent": 944.8060075093867,
              },
              "size": Object {
                "absolute": 7549,
                "ok": true,
                "percent": 944.8060075093867,
              },
            },
            "entry": "one",
            "rule": Object {
              "maxAsyncSizeDiff": Infinity,
              "maxInitialSizeDiff": 100,
              "maxSizeDiff": Infinity,
            },
            "useCompressedSize": false,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              before: #.before,
              after: #.after,
              diff: #.diff,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"one\\": Download size diff of initial assets is 7.37 kb (944.81%). It's over the 0.10 kb limit",
    "related": Array [
      Object {
        "id": "one",
        "type": "entry",
      },
    ],
  },
  Object {
    "compilation": "9eb9a218218f1ee4ba40",
    "details": Array [
      Object {
        "payload": Object {
          "context": Object {
            "after": Object {
              "asyncSize": 1200,
              "initialSize": 29129,
              "size": 30329,
            },
            "before": Object {
              "asyncSize": 142,
              "initialSize": 12277,
              "size": 12419,
            },
            "compilation": "9eb9a218218f1ee4ba40",
            "diff": Object {
              "asyncSize": Object {
                "absolute": 1058,
                "ok": true,
                "percent": 745.0704225352113,
              },
              "initialSize": Object {
                "absolute": 16852,
                "ok": false,
                "percent": 137.26480410523743,
              },
              "size": Object {
                "absolute": 17910,
                "ok": true,
                "percent": 144.21451002496175,
              },
            },
            "entry": "two",
            "rule": Object {
              "maxAsyncSizeDiff": Infinity,
              "maxInitialSizeDiff": 100,
              "maxSizeDiff": Infinity,
            },
            "useCompressedSize": false,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              before: #.before,
              after: #.after,
              diff: #.diff,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"two\\": Download size diff of initial assets is 16.46 kb (137.26%). It's over the 0.10 kb limit",
    "related": Array [
      Object {
        "id": "two",
        "type": "entry",
      },
    ],
  },
]
`;

exports[`maxSizeDiff global 1`] = `
Array [
  Object {
    "compilation": "9eb9a218218f1ee4ba40",
    "details": Array [
      Object {
        "payload": Object {
          "context": Object {
            "after": Object {
              "asyncSize": 0,
              "initialSize": 1887,
              "size": 1887,
            },
            "before": Object {
              "asyncSize": 0,
              "initialSize": 449,
              "size": 449,
            },
            "compilation": "9eb9a218218f1ee4ba40",
            "diff": Object {
              "asyncSize": Object {
                "absolute": 0,
                "ok": true,
                "percent": 0,
              },
              "initialSize": Object {
                "absolute": 1438,
                "ok": true,
                "percent": 320.26726057906455,
              },
              "size": Object {
                "absolute": 1438,
                "ok": false,
                "percent": 320.26726057906455,
              },
            },
            "entry": "one",
            "rule": Object {
              "maxAsyncSizeDiff": Infinity,
              "maxInitialSizeDiff": Infinity,
              "maxSizeDiff": 100,
            },
            "useCompressedSize": true,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              before: #.before,
              after: #.after,
              diff: #.diff,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"one\\": Download size diff of assets is 1.40 kb (320.27%). It's over the 0.10 kb limit",
    "related": Array [
      Object {
        "id": "one",
        "type": "entry",
      },
    ],
  },
  Object {
    "compilation": "9eb9a218218f1ee4ba40",
    "details": Array [
      Object {
        "payload": Object {
          "context": Object {
            "after": Object {
              "asyncSize": 590,
              "initialSize": 12211,
              "size": 12801,
            },
            "before": Object {
              "asyncSize": 134,
              "initialSize": 8615,
              "size": 8749,
            },
            "compilation": "9eb9a218218f1ee4ba40",
            "diff": Object {
              "asyncSize": Object {
                "absolute": 456,
                "ok": true,
                "percent": 340.2985074626866,
              },
              "initialSize": Object {
                "absolute": 3596,
                "ok": true,
                "percent": 41.74114915844458,
              },
              "size": Object {
                "absolute": 4052,
                "ok": false,
                "percent": 46.313864441650466,
              },
            },
            "entry": "two",
            "rule": Object {
              "maxAsyncSizeDiff": Infinity,
              "maxInitialSizeDiff": Infinity,
              "maxSizeDiff": 100,
            },
            "useCompressedSize": true,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              before: #.before,
              after: #.after,
              diff: #.diff,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"two\\": Download size diff of assets is 3.96 kb (46.31%). It's over the 0.10 kb limit",
    "related": Array [
      Object {
        "id": "two",
        "type": "entry",
      },
    ],
  },
]
`;

exports[`maxSizeDiff not match - global 1`] = `Array []`;

exports[`maxSizeDiff not match - override global 1`] = `Array []`;

exports[`maxSizeDiff override global 1`] = `
Array [
  Object {
    "compilation": "9eb9a218218f1ee4ba40",
    "details": Array [
      Object {
        "payload": Object {
          "context": Object {
            "after": Object {
              "asyncSize": 0,
              "initialSize": 1887,
              "size": 1887,
            },
            "before": Object {
              "asyncSize": 0,
              "initialSize": 449,
              "size": 449,
            },
            "compilation": "9eb9a218218f1ee4ba40",
            "diff": Object {
              "asyncSize": Object {
                "absolute": 0,
                "ok": true,
                "percent": 0,
              },
              "initialSize": Object {
                "absolute": 1438,
                "ok": true,
                "percent": 320.26726057906455,
              },
              "size": Object {
                "absolute": 1438,
                "ok": false,
                "percent": 320.26726057906455,
              },
            },
            "entry": "one",
            "rule": Object {
              "maxAsyncSizeDiff": Infinity,
              "maxInitialSizeDiff": Infinity,
              "maxSizeDiff": 100,
            },
            "useCompressedSize": true,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              before: #.before,
              after: #.after,
              diff: #.diff,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"one\\": Download size diff of assets is 1.40 kb (320.27%). It's over the 0.10 kb limit",
    "related": Array [
      Object {
        "id": "one",
        "type": "entry",
      },
    ],
  },
  Object {
    "compilation": "9eb9a218218f1ee4ba40",
    "details": Array [
      Object {
        "payload": Object {
          "context": Object {
            "after": Object {
              "asyncSize": 590,
              "initialSize": 12211,
              "size": 12801,
            },
            "before": Object {
              "asyncSize": 134,
              "initialSize": 8615,
              "size": 8749,
            },
            "compilation": "9eb9a218218f1ee4ba40",
            "diff": Object {
              "asyncSize": Object {
                "absolute": 456,
                "ok": true,
                "percent": 340.2985074626866,
              },
              "initialSize": Object {
                "absolute": 3596,
                "ok": true,
                "percent": 41.74114915844458,
              },
              "size": Object {
                "absolute": 4052,
                "ok": false,
                "percent": 46.313864441650466,
              },
            },
            "entry": "two",
            "rule": Object {
              "maxAsyncSizeDiff": Infinity,
              "maxInitialSizeDiff": Infinity,
              "maxSizeDiff": 100,
            },
            "useCompressedSize": true,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              before: #.before,
              after: #.after,
              diff: #.diff,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"two\\": Download size diff of assets is 3.96 kb (46.31%). It's over the 0.10 kb limit",
    "related": Array [
      Object {
        "id": "two",
        "type": "entry",
      },
    ],
  },
]
`;

exports[`maxSizeDiff with useCompressedSize = false 1`] = `
Array [
  Object {
    "compilation": "9eb9a218218f1ee4ba40",
    "details": Array [
      Object {
        "payload": Object {
          "context": Object {
            "after": Object {
              "asyncSize": 0,
              "initialSize": 8348,
              "size": 8348,
            },
            "before": Object {
              "asyncSize": 0,
              "initialSize": 799,
              "size": 799,
            },
            "compilation": "9eb9a218218f1ee4ba40",
            "diff": Object {
              "asyncSize": Object {
                "absolute": 0,
                "ok": true,
                "percent": 0,
              },
              "initialSize": Object {
                "absolute": 7549,
                "ok": true,
                "percent": 944.8060075093867,
              },
              "size": Object {
                "absolute": 7549,
                "ok": false,
                "percent": 944.8060075093867,
              },
            },
            "entry": "one",
            "rule": Object {
              "maxAsyncSizeDiff": Infinity,
              "maxInitialSizeDiff": Infinity,
              "maxSizeDiff": 100,
            },
            "useCompressedSize": false,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              before: #.before,
              after: #.after,
              diff: #.diff,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"one\\": Download size diff of assets is 7.37 kb (944.81%). It's over the 0.10 kb limit",
    "related": Array [
      Object {
        "id": "one",
        "type": "entry",
      },
    ],
  },
  Object {
    "compilation": "9eb9a218218f1ee4ba40",
    "details": Array [
      Object {
        "payload": Object {
          "context": Object {
            "after": Object {
              "asyncSize": 1200,
              "initialSize": 29129,
              "size": 30329,
            },
            "before": Object {
              "asyncSize": 142,
              "initialSize": 12277,
              "size": 12419,
            },
            "compilation": "9eb9a218218f1ee4ba40",
            "diff": Object {
              "asyncSize": Object {
                "absolute": 1058,
                "ok": true,
                "percent": 745.0704225352113,
              },
              "initialSize": Object {
                "absolute": 16852,
                "ok": true,
                "percent": 137.26480410523743,
              },
              "size": Object {
                "absolute": 17910,
                "ok": false,
                "percent": 144.21451002496175,
              },
            },
            "entry": "two",
            "rule": Object {
              "maxAsyncSizeDiff": Infinity,
              "maxInitialSizeDiff": Infinity,
              "maxSizeDiff": 100,
            },
            "useCompressedSize": false,
          },
        },
        "query": "
            $input: resolveInputFile();
            {
              entry: #.entry.resolveEntrypoint(#.compilation),
              useCompressedSize: #.useCompressedSize,
              before: #.before,
              after: #.after,
              diff: #.diff,
              rule: #.rule,
            }
            ",
        "type": "discovery",
      },
    ],
    "filename": "input.json",
    "message": "Entry \\"two\\": Download size diff of assets is 17.49 kb (144.21%). It's over the 0.10 kb limit",
    "related": Array [
      Object {
        "id": "two",
        "type": "entry",
      },
    ],
  },
]
`;
